### DISCUSSION TITLE
Set recurring jobs to PVCs

### DISCUSSION THREAD

**Discussion by cod-r** [longhorn/longhorn#5773] [Q&A]
I have a recurring job that I want to apply to some PVCs, but not to all of them.
I'm creating my PVCs using Argo CD. 

From the official documentation it says that i should add a label to the PV to create backups:
```
kubectl -n longhorn-system label volume/pvc-8b9cd514-4572-4eb2-836a-ed311e804d2f recurring-job.longhorn.io/backup1=enabled
```
but the PV is created dynamically by longhorn. Because I'm using gitops and the configuration stays in git, I can't modify the PVs using kubectl.

Why can't we set the label to the PVC considering the PVs are created dynamically?
Is there any way to do what i'm looking for?

**Comment by c3y1huang** [comment]
Have you considered using a new [Longhorn StorageClass that includes RecurringJob assignment](https://longhorn.io/docs/1.4.1/snapshots-and-backups/scheduling-backups-and-snapshots/#recurring-jobs-assignment-using-a-storageclass) when creating those PVCs?

**Reply by cod-r** [reply]
I already have lots of PVCs created with the default `longhorn` storageclass. How would I transfer those PVCs to the new storage class? What if I want to stop the backups for some volumes?

Just adding or removing a label to the PVCs would be really convenient. But I don't understand this limitation of adding the label to the PVs. Most of the PVs are created dinamically.

My question is why can't I add the label to the PVCs?

**Reply by c3y1huang** [reply]
Recurring job processes are created on a per-labeled volume basis. It makes sense for Longhorn to allow selecting the recurring jobs at the StorageClass level because this enables the CSI to apply the labels to all volumes created by that StorageClass. Supporting recurring jobs at the PVC level would require Longhorn to handle each PVC resource change, resulting in unnecessary resource overhead. Therefore, setting the StorageClass or labeling the volume is a more straightforward and efficient solution.

**Comment by cod-r** [comment]
_"Supporting recurring jobs at the PVC level would require Longhorn to handle each PVC resource change, resulting in unnecessary resource overhead."_
From my understanding you already do this with PVs instead of PVCs. How is this different on resource overhead?

_"Therefore, setting the StorageClass or labeling the volume is a more straightforward and efficient solution"._
I understand the solution with having two storage classes, one with backups and one without. And this can be configured with gitops. 
But this is not covering the use case when you want to stop the backups for an existing volume, or to add backups to an existing volume. This would need a transfer for a volume from a storage class to another. And this is not a simple change.

And labeling the PVs is not compatible with the gitops approach, because I'm not creating any PVs directly, I'm only creating the PVCs.

Should I create an issue to request this feature? I'm sure more people doing gitops would need it.

**Reply by c3y1huang** [reply]
Longhorn supports assigning the recurring job to the volumes, which is different from the Kubernetes PV and PVCs. Having supporting recurring jobs at the PVC means Longhorn needs to have a controller to watch over the PVC resource changes.

Have you also considered using recurring job groups? You can modify the recurring job within the group and the changes will be effective for the existing volumes.

**Comment by c3y1huang** [comment]
Feature request created: https://github.com/longhorn/longhorn/issues/5791
